op-program/client/interop/interop.go

```diff
@@ -0,0 +1,118 @@
+package interop
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum-optimism/optimism/op-program/client/boot"
+	"github.com/ethereum-optimism/optimism/op-program/client/claim"
+	"github.com/ethereum-optimism/optimism/op-program/client/interop/types"
+	"github.com/ethereum-optimism/optimism/op-program/client/l1"
+	"github.com/ethereum-optimism/optimism/op-program/client/l2"
+	"github.com/ethereum-optimism/optimism/op-program/client/tasks"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+var (
+	ErrIncorrectOutputRootType = errors.New("incorrect output root type")
+)
+
+type taskExecutor interface {
+	RunDerivation(
+		logger log.Logger,
+		rollupCfg *rollup.Config,
+		l2ChainConfig *params.ChainConfig,
+		l1Head common.Hash,
+		agreedOutputRoot eth.Bytes32,
+		claimedBlockNumber uint64,
+		l1Oracle l1.Oracle,
+		l2Oracle l2.Oracle) (tasks.DerivationResult, error)
+}
+
+func RunInteropProgram(logger log.Logger, bootInfo *boot.BootInfo, l1PreimageOracle l1.Oracle, l2PreimageOracle l2.Oracle, validate bool) error {
+	return runInteropProgram(logger, bootInfo, l1PreimageOracle, l2PreimageOracle, validate, &interopTaskExecutor{})
+}
+
+func runInteropProgram(logger log.Logger, bootInfo *boot.BootInfo, l1PreimageOracle l1.Oracle, l2PreimageOracle l2.Oracle, validate bool, tasks taskExecutor) error {
+	logger.Info("Interop Program Bootstrapped", "bootInfo", bootInfo)
+
+	// For the first step in a timestamp, we would get a SuperRoot as the agreed claim - TransitionStateByRoot will
+	// automatically convert it to a TransitionState with Step: 0.
+	transitionState := l2PreimageOracle.TransitionStateByRoot(bootInfo.L2OutputRoot)
+	if transitionState.Version() != types.IntermediateTransitionVersion {
+		return fmt.Errorf("%w: %v", ErrIncorrectOutputRootType, transitionState.Version())
+	}
+
+	super, err := eth.UnmarshalSuperRoot(transitionState.SuperRoot)
+	if err != nil {
+		return fmt.Errorf("invalid super root: %w", err)
+	}
+	if super.Version() != eth.SuperRootVersionV1 {
+		return fmt.Errorf("%w: %v", ErrIncorrectOutputRootType, super.Version())
+	}
+	superRoot := super.(*eth.SuperV1)
+	claimedBlockNumber, err := bootInfo.RollupConfig.TargetBlockNumber(superRoot.Timestamp + 1)
+	if err != nil {
+		return err
+	}
+	derivationResult, err := tasks.RunDerivation(
+		logger,
+		bootInfo.RollupConfig,
+		bootInfo.L2ChainConfig,
+		bootInfo.L1Head,
+		superRoot.Chains[0].Output,
```

Comment:
I know the hardcoded zero here is for testing the first step. But just so we don't forget: with the ChainIDs/output tuple, you can lookup the output from the rollup chainID.

Reply:
Switched this to use the Step from the agreed transition state. We'll need to handle no-op steps where Step is greater than the length of the array but we have action tests for that we'll work towards later anyway and that's better than a hard coded 0.

Reply:
Looks like it's still hardcoded to 0.

Reply:
nvmind. I see this is done in a different PR in the stack.